### PR TITLE
tokenizing input: added class Tokenizer, tests, changed handleCommand, fix wildcards, spacing

### PR DESCRIPTION
**Pull request by willyjlee**
Added Tokenizer class, Tokenizer tests, handleCommand, and List instead of Scanner for commands

### CODE REVIEW

**Comment by nikkibisarya** [comment]
LGTM

File: src/codeu/chat/client/commandline/Chat.java [hunk 1]
```diff
@@ -14,15 +14,14 @@
 
 package codeu.chat.client.commandline;
 
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Scanner;
-import java.util.Stack;
+import java.util.*;
```
Comment (adrnswanberg) [inline src/codeu/chat/client/commandline/Chat.java]: Wildcard imports violates the Google java style guide. Please revert back to importing individual packages.

File: src/codeu/chat/util/Tokenizer.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
+package codeu.chat.util;
+
+import java.io.IOException;
+import java.lang.*;
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: Remove wildcard import.

File: test/codeu/chat/util/TokenizerTest.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package codeu.chat.util;
+
+import java.io.IOException;
+import static org.junit.Assert.*;
```
Comment (adrnswanberg) [inline test/codeu/chat/util/TokenizerTest.java]: Remove wildcard import.

File: src/codeu/chat/util/Tokenizer.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+public final class Tokenizer{
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: add space before open bracket, here and everywhere else.

File: src/codeu/chat/util/Tokenizer.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    token.setLength(0);
+    while(remaining()>0 && !Character.isWhitespace(peek())){
+      token.append(read());
+    }
+    return token.toString();
+  }
+
+  private String readWithQuotes() throws IOException{
+    token.setLength(0);
+    if(read()!='"'){
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: add space between symbols, here and below

File: src/codeu/chat/client/commandline/Chat.java [hunk 6]
```diff
@@ -45,11 +44,15 @@ public Chat(Context context) {
...
   //
-  public boolean handleCommand(String line) {
+  public boolean handleCommand(String line) throws IOException {
 
-    final Scanner tokens = new Scanner(line.trim());
-
-    final String command = tokens.hasNext() ? tokens.next() : "";
+    final List<String> args = new ArrayList<>();
+    final Tokenizer tokenizer = new Tokenizer(line);
+    for (String token = tokenizer.next(); token!=null; token = tokenizer.next()){
```
Comment (adrnswanberg) [inline src/codeu/chat/client/commandline/Chat.java]: add spaces to this expression: `token!=null;`

File: src/codeu/chat/client/commandline/Chat.java [hunk 7]
```diff
@@ -136,8 +139,8 @@ public void invoke(Scanner args) {
     //
     panel.register("u-add", new Panel.Command() {
       @Override
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        final String name = args.size()>0 ? args.get(0).trim() : ""; // this only takes first token of the name (surround with quotes for inputs with whitespaces)
```
Comment (adrnswanberg) [inline src/codeu/chat/client/commandline/Chat.java]: nit: add symbols between spaces, and move comment to previous line.

File: src/codeu/chat/client/commandline/Chat.java [hunk 8]
```diff
@@ -155,8 +158,8 @@ public void invoke(Scanner args) {
     //
     panel.register("u-sign-in", new Panel.Command() {
       @Override
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        final String name = args.size()>0 ? args.get(0).trim() : ""; // this only takes first token of the name (surround with quotes for inputs with whitespaces)
```
Comment (adrnswanberg) [inline src/codeu/chat/client/commandline/Chat.java]: same here.

File: src/codeu/chat/client/commandline/Chat.java [hunk 9]
```diff
@@ -238,8 +241,8 @@ public void invoke(Scanner args) {
     //
     panel.register("c-add", new Panel.Command() {
       @Override
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        final String name = args.size()>0 ? args.get(0).trim() : ""; // this only takes first token of the conversation name (surround with quotes for inputs with whitespaces)
```
Comment (adrnswanberg) [inline src/codeu/chat/client/commandline/Chat.java]: same here

File: src/codeu/chat/client/commandline/Chat.java [hunk 10]
```diff
@@ -260,8 +263,8 @@ public void invoke(Scanner args) {
     //
     panel.register("c-join", new Panel.Command() {
       @Override
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        final String name = args.size()>0 ? args.get(0).trim() : ""; // this only takes first token of the conversation name (surround with quotes for inputs with whitespaces)
```
Comment (adrnswanberg) [inline src/codeu/chat/client/commandline/Chat.java]: same here

File: src/codeu/chat/client/commandline/Chat.java [hunk 11]
```diff
@@ -361,8 +364,8 @@ public void invoke(Scanner args) {
     //
     panel.register("m-add", new Panel.Command() {
       @Override
-      public void invoke(Scanner args) {
-        final String message = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+        final String message = args.size()>0 ? args.get(0).trim() : ""; // this only takes first token of the msg (surround with quotes for inputs with whitespaces)
```
Comment (adrnswanberg) [inline src/codeu/chat/client/commandline/Chat.java]: same here

File: src/codeu/chat/client/commandline/Panel.java [hunk 12]
```diff
@@ -14,9 +14,7 @@
 
 package codeu.chat.client.commandline;
 
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Scanner;
+import java.util.*;
```
Comment (adrnswanberg) [inline src/codeu/chat/client/commandline/Panel.java]: remove wildcard import

File: src/codeu/chat/client/commandline/Panel.java [hunk 13]
```diff
@@ -48,10 +46,10 @@ public void register(String commandName, Command command) {
   // will be returned. True will be return if a command is found. Whether or not
   // the command was successful is not returned.
   //
-  public boolean handleCommand(String commandName, Scanner line) {
+  public boolean handleCommand(String commandName, List<String>args) {
     final Command command = commands.get(commandName);
-    if (command != null) {
-      command.invoke(line);
+    if(command != null){
```
Comment (adrnswanberg) [inline src/codeu/chat/client/commandline/Panel.java]: re-insert spaces between symbols

File: src/codeu/chat/util/Tokenizer.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  public String next() throws IOException{
+    while(remaining()>0 && Character.isWhitespace(peek())){
+      read();
+    }
+    if(remaining()<=0){
+      return null;
+    }
+    else if(peek() == '"'){
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: move `else if` to start on previous line

File: src/codeu/chat/util/Tokenizer.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+      return readWithQuotes();
+    }
+    else{
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: same here

**Review by adrnswanberg — CHANGES_REQUESTED** [review]

File: src/codeu/chat/util/Tokenizer.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    at = 0;
+    token = new StringBuilder();
+  }
+
+  private int remaining() {
+    return source.length() - at;
+  }
+
+  private char peek() throws IOException {
+    if(at < source.length()) {
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: add space after `if`

File: src/codeu/chat/util/Tokenizer.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+  private char read() throws IOException {
+    final char c = peek();
+    at += 1;
+    return c;
+  }
+
+  private String readWithNoQuotes() throws IOException {
+    token.setLength(0);
+    while(remaining() > 0 && !Character.isWhitespace(peek())){
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: add space `while`

File: src/codeu/chat/util/Tokenizer.java [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+      token.append(read());
+    }
+    return token.toString();
+  }
+
+  private String readWithQuotes() throws IOException {
+    token.setLength(0);
+    if(read() != '"') {
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: add space after if

File: src/codeu/chat/util/Tokenizer.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+      throw new IOException("String must start with open quote");
+    }
+    while(peek() != '"') {
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: add space after while

File: src/codeu/chat/util/Tokenizer.java [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+      token.append(read());
+    }
+    read();
+    return token.toString();
+  }
+
+  public String next() throws IOException {
+    while(remaining() > 0 && Character.isWhitespace(peek())){
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: add space after while

File: src/codeu/chat/util/Tokenizer.java [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+      read();
+    }
+    if(remaining() <= 0) {
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: add space after if

File: src/codeu/chat/util/Tokenizer.java [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+      return null;
+    } else if(peek() == '"'){
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: add space after if

File: src/codeu/chat/util/Tokenizer.java [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+      return readWithQuotes();
+    } else{
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Tokenizer.java]: nit: add space after else

**Review by adrnswanberg — APPROVED** [review]
Nice! Just address a few nits and you're free to merge.

**Pull request merged**